### DISCUSSION TITLE
orm.reconstructor vs __post_init__

### DISCUSSION THREAD

**Discussion by zeehio** [sqlalchemy/sqlalchemy#11794] [Ideas]
As documented in https://docs.sqlalchemy.org/en/13/orm/constructors.html, the ORM does not call `__init__` when creating new objects. If some init-like operation has to happen, the example in that page shows how to use the `reconstructor` decorator:

```python
from sqlalchemy import orm

class MyMappedClass(object):
    def __init__(self, data):
        self.data = data
        # we need stuff on all instances, but not in the database.
        self.stuff = []

    @orm.reconstructor
    def init_on_load(self):
        self.stuff = []
```

[PEP557:  Data Classes](https://peps.python.org/pep-0557/#post-init-processing) introduced dataclasses. In that PEP, in the scope of dataclasses, the authors introduce the `__post_init__` method, for doing additional initializations, similar to sqlalchemy reconstructor's decorator.

I feel it would be easy to adopt the `__post_init__` convention in sqlalchemy, instead of having a decorated method.

The easiest implementation I can think of would be modifying this piece of code:

https://github.com/sqlalchemy/sqlalchemy/blob/5b117f3d4b38d12d61a39fc60582d4348232334f/lib/sqlalchemy/orm/mapper.py#L1529-L1540

So it looks like this:

```python
 for key, method in util.iterate_attributes(self.class_): 
     if key == "__init__" and hasattr(method, "_sa_original_init"): 
         method = method._sa_original_init 
         if hasattr(method, "__func__"): 
             method = method.__func__ 
     # change starts here (adding these lines)
     if key == "__post_init__":
         self._reconstructor = method 
         event.listen(manager, "load", _event_on_load, raw=True)
     # change ends here
     if callable(method): 
         if hasattr(method, "__sa_reconstructor__"): 
             self._reconstructor = method 
             event.listen(manager, "load", _event_on_load, raw=True) 
         elif hasattr(method, "__sa_validators__"): 
             validation_opts = method.__sa_validation_opts__ 
             for name in method.__sa_validators__: 
```

The change seems fairly simple, but I do not know if all the involved work (evaluating if reconstructor should be deprecated, handling corner cases where someone has written a class that uses reconstructor but also may have a `__post_init__` method, documentation updates... etc) is worth the effort. 

I'd rather ask for feedback, opinions and guidance before considering starting an implementation.

Thanks for reading!

**Comment by CaselIT** [comment]
Hi,

SQLAlchemy supports mapping dataclasses, so there may be a bit of an overlap there, but as long as the current decorator takes precedence (so that people using dataclasses would have the opinion of using two methods) I think it would be ok.

Not sure if we need to provide a way of option out of that naming convention for some reason. Probably not, since it that the case an empty method could be decorated.

@zzzeek what's your opinion on this?

**Comment by zzzeek** [comment]
> the authors introduce the __post_init__ method, for doing additional initializations, similar to sqlalchemy reconstructor's decorator.

These aren't similar at all IMO.  `__post_init__` is a dataclass method that's associated with the `__init__` phase of object construction.   "reconstructor" is specific to SQLAlchemy's object deserialization which is an entirely different lifecycle.   it can't be assumed that someone's `__post_init__` performs only tasks that are appropriate and necessary within the ORM object loading phase.

> I feel it would be easy to adopt the __post_init__ convention in sqlalchemy, instead of having a decorated method.

If you want to re-use your `__post_init__` implementation in the reconstructor workflow, make a mixin or base class something like this:

```py
class ReconstuctorPostInit:
     @reconstructor
     def _reconstruct(self):
         self.__post_init__()
```

It would be inappropriate for this to be a default behavior as `__post_init__` has nothing to do with ORMs and it's a large leap to assume this method is exactly the implementation for what the ORMs object deserialization would be doing.

>  (evaluating if reconstructor should be deprecated, 

This would be inappropriate.

1. it would be extremely surprising for `__post_init__` to be called everytime objects are loaded by default.   These two hooks have entirely different semantics
2. usage of SQLAlchemy ORM without using dataclasses is not deprecated and it would be confusing for dataclass-oriented naming conventions to slowly leak into the ORM in an inconsistent way

**Comment by zeehio** [comment]
Thanks for your feedback.

I feel sqlalchemy is extremely versatile and also complex. I feel there are so many ways to do things, and all of them make sense for some use case. It's hard to understand how everything works.

After reading your reply it is clear to me that my proposal was wrong and that I do not have enough background on sqlalchemy for making these kind of naive suggestions. I hope to learn more and eventually be able to make a useful contribution.

Thanks again for your feedback, I will now close this discussion.

**Reply by zzzeek** [reply]
Please dont feel naive about it, we are extremely conservative about changing things after many years of rushing into things too quickly.  thanks
